### PR TITLE
feat: duplicate block commands

### PR DESCRIPTION
**Pull request by Kneesal**
# Description

### Issue

a custom hook the allows the undo/redo of a duplicate step and block

[Link to Basecamp Todo](https://3.basecamp.com/3105655/projects)

### CODE REVIEW

File: apps/journeys-admin/src/components/Editor/Slider/Content/Canvas/QuickControls/DuplicateBlock/DuplicateBlock.tsx [hunk 1]
```diff
@@ -30,57 +50,110 @@ export function DuplicateBlock({
...
-      if (parentOrder == null) return
+  function handleDuplicateBlock(): void {
+    if (
+      selectedBlock == null ||
+      journey == null ||
+      selectedBlock?.parentOrder == null
+    )
+      return
+    const { parentOrder } = selectedBlock
+    if (parentOrder == null) return
```
Comment (Kneesal) [inline apps/journeys-admin/src/components/Editor/Slider/Content/Canvas/QuickControls/DuplicateBlock/DuplicateBlock.tsx:83]: moving this above does not infer the type properly

File: apps/journeys-admin/src/components/Editor/utils/useBlockDuplicateCommand/useBlockDuplicateCommand.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+  function addBlockDuplicate({
+    block,
+    execute
+  }: AddBlockDuplicateParams): void {
+    add({
+      parameters: { execute: {}, undo: {} },
+      execute() {
+        execute()
+      },
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/utils/useBlockDuplicateCommand/useBlockDuplicateCommand.tsx:40]:
```suggestion
      execute,
```

File: apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/nodes/StepBlockNode/StepBlockNodeMenu/DuplicateStep/DuplicateStep.tsx [hunk 3]
```diff
@@ -54,84 +72,136 @@ export function DuplicateStep({
...
-                  `
-                })
-                return [...existingBlockRefs, duplicatedBlockRef]
-              }
-            }
-          })
+    const idMap: BlockDuplicateIdMap[] = [
+      { oldId: step.id, newId: stepBlock.id }
+    ]
+    const stepChildren = flatten(step.children)?.map((block) => {
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/Slider/JourneyFlow/nodes/StepBlockNode/StepBlockNodeMenu/DuplicateStep/DuplicateStep.tsx:109]:
```suggestion
    const stepDescendants = flatten(step.children)?.map((block) => {
```

File: apps/journeys-admin/src/components/Editor/utils/useBlockDuplicateCommand/useBlockDuplicateCommand.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            })
+          : dispatch({
+              type: 'SetEditorFocusAction',
+              selectedStepId: selectedStep?.id,
+              selectedBlockId: selectedBlock?.id,
+              activeSlide: ActiveSlide.Content,
+              activeContent: ActiveContent.Canvas
+            })
+        void blockDelete(block, {
+          optimisticResponse: { blockDelete: [] }
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/utils/useBlockDuplicateCommand/useBlockDuplicateCommand.tsx:59]: not the right response.

File: apps/journeys-admin/src/components/Editor/utils/useBlockDuplicateCommand/useBlockDuplicateCommand.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+              selectedStepId: selectedStep?.id,
+              selectedBlockId: block?.id,
+              activeSlide: ActiveSlide.Content,
+              activeContent: ActiveContent.Canvas
+            })
+
+        void blockRestore({
+          variables: { id: block.id },
+          optimisticResponse: {
+            blockRestore: [block as BlockRestore]
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/utils/useBlockDuplicateCommand/useBlockDuplicateCommand.tsx:82]: not the right response.

**Review by tataihono — CHANGES_REQUESTED** [review]

File: apps/journeys-admin/src/components/Editor/utils/useBlockDuplicateCommand/useBlockDuplicateCommand.tsx [hunk 6]
```diff
@@ -24,22 +26,35 @@ export function useBlockDuplicateCommand(): {
...
+      result.push(...flatten(item.children))
+      return result
+    }, [])
+  }
+
   function addBlockDuplicate({
     block,
     execute
   }: AddBlockDuplicateParams): void {
+    if (selectedStep == null || journey == null || steps == null) return
```
Comment (tataihono) [inline apps/journeys-admin/src/components/Editor/utils/useBlockDuplicateCommand/useBlockDuplicateCommand.tsx:46]: do you even need journey in this?

**Review by tataihono — CHANGES_REQUESTED** [review]

**Review by tataihono — APPROVED** [review]

**Pull request merged**